iroh-relay/src/protos/relay.rs

```diff
@@ -157,6 +157,47 @@ impl<T: AsRef<[u8]>> From<T> for Datagrams {
 }
 
 impl Datagrams {
+    /// Splits the current datagram into at maximum `num_segments` segments.
+    ///
+    /// Calling this on a datagram batch that only contains a single datagram (`segment_size == None`)
+    /// will result in returning essentially `Some(self.clone())`, while making `self` empty afterwards.
+    ///
+    /// Calling this on a datagram batch with e.g. 15 datagrams with `num_segments == 10` will
+    /// result in returning `Some(datagram_batch)` where that `datagram_batch` contains 10 datagrams and
+    /// `self` contains 5 datagrams.
+    ///
+    /// Calling this on a datagram batch that doesn't contain `num_segments` datagrams, but less
+    /// will result in making `self` empty and returning essentially a clone of `self`.
+    ///
+    /// Calling this on an empty datagram batch (i.e. one where `contents.is_empty()`) will return `None`.
+    pub fn take_segments(&mut self, num_segments: usize) -> Option<Datagrams> {
+        if self.contents.is_empty() {
+            return None;
+        }
+
+        let Some(segment_size) = self.segment_size else {
+            let contents = std::mem::take(&mut self.contents);
+            return Some(Datagrams {
+                ecn: self.ecn,
+                segment_size: None,
+                contents,
+            });
+        };
+
+        let usize_segment_size = usize::from(u16::from(segment_size));
+        let max_content_len = num_segments * usize_segment_size;
+        let contents = self
+            .contents
+            .split_to(std::cmp::min(max_content_len, self.contents.len()));
+
+        Some(Datagrams {
+            ecn: self.ecn,
+            segment_size: (num_segments > 1 && usize_segment_size < contents.len())
```

Comment:
this is very hard to read

Reply:
Would this be better?
```rs
let is_datagram_batch = num_segments > 1 && usize_segment_size < contents.len();

Some(Datagrams {
    ecn: self.ecn,
    segment_size: is_datagram_batch.then_some(segment_size),
    contents,
})
```

Reply:
Or would you prefer if I rewrote this using if-else?

Reply:
the split out seems fine